buildSrc/src/main/groovy/MuzzlePlugin.groovy

```diff
@@ -244,47 +239,92 @@ class MuzzlePlugin implements Plugin<Project> {
   private static Set<MuzzleDirective> inverseOf(MuzzleDirective muzzleDirective, RepositorySystem system, RepositorySystemSession session) {
     Set<MuzzleDirective> inverseDirectives = new HashSet<>()
 
-    Artifact allVerisonsArtifact = new DefaultArtifact(muzzleDirective.group, muzzleDirective.module, "jar", "[,)")
+    Artifact allVersionsArtifact = new DefaultArtifact(muzzleDirective.group, muzzleDirective.module, "jar", "[,)")
     Artifact directiveArtifact = new DefaultArtifact(muzzleDirective.group, muzzleDirective.module, "jar", muzzleDirective.versions)
 
-
     VersionRangeRequest allRangeRequest = new VersionRangeRequest()
     allRangeRequest.setRepositories(MUZZLE_REPOS)
-    allRangeRequest.setArtifact(allVerisonsArtifact)
+    allRangeRequest.setArtifact(allVersionsArtifact)
     VersionRangeResult allRangeResult = system.resolveVersionRange(session, allRangeRequest)
 
     VersionRangeRequest rangeRequest = new VersionRangeRequest()
     rangeRequest.setRepositories(MUZZLE_REPOS)
     rangeRequest.setArtifact(directiveArtifact)
     VersionRangeResult rangeResult = system.resolveVersionRange(session, rangeRequest)
-    Set<Version> versions = rangeResult.versions.toSet()
-
-    filterVersion(allRangeResult.versions.toSet(), muzzleDirective.skipVersions).collect { version ->
-      if (!versions.contains(version)) {
-        MuzzleDirective inverseDirective = new MuzzleDirective()
-        inverseDirective.group = muzzleDirective.group
-        inverseDirective.module = muzzleDirective.module
-        inverseDirective.versions = "$version"
-        inverseDirective.assertPass = !muzzleDirective.assertPass
-        inverseDirectives.add(inverseDirective)
-      }
+
+    allRangeResult.getVersions().removeAll(rangeResult.getVersions())
+
+    filterVersions(allRangeResult, muzzleDirective.skipVersions).each { version ->
+      MuzzleDirective inverseDirective = new MuzzleDirective()
+      inverseDirective.group = muzzleDirective.group
+      inverseDirective.module = muzzleDirective.module
+      inverseDirective.versions = version
+      inverseDirective.assertPass = !muzzleDirective.assertPass
+      inverseDirectives.add(inverseDirective)
     }
 
     return inverseDirectives
   }
 
-  private static void limitLargeRanges(VersionRangeResult result, Set<Version> versions, Set<String> skipVersions) {
-    List<Version> copy = new ArrayList<>(versions)
-    copy.removeAll(skipVersions)
+  private static Set<String> filterVersions(VersionRangeResult range, Set<String> skipVersions) {
+    Set<String> result = new HashSet<>()
+
+    def predicate = new AcceptableVersions(range, skipVersions)
+    if (predicate.test(range.lowestVersion)) {
+      result.add(range.lowestVersion.toString())
+    }
+    if (predicate.test(range.highestVersion)) {
+      result.add(range.highestVersion.toString())
+    }
```

Comment:
I think maybe good not to apply predicate test to `highestVersion`, so we can find out early that we may issue with future version.

Though then we tend to just limit the max version, and we never get notified when a non-"beta" version is released, which is worse.

So ignore me, all good 😄.